Use `xcode-select -p` (dlsym() approach)

norio-nomura opened the pull request:
Another approach for fixing #167

Improved from #172:
- No `execv`
- SPM compatible
- Searching toolchain path is extended by applying reviews on #172.

Progress:
- [x] sourcekitd.framework
- [x] libclang.dylib

norio-nomura commented:
### Additional notes
#### library search order:

1. `XCODE_DEFAULT_TOOLCHAIN_OVERRIDE/usr/lib` # for debugging on Xcode launched by `launch-with-toolchain`
2. `$TOOLCHAIN_DIR/usr/lib` # for debugging on Xcode
3. \`xcode-select -p\` `/Toolchains/XcodeDefault.xctoolchain/usr/lib`
4. `(Application dir)/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib`
5. `(Application dir)/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib`
6. `(User application dir)/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib`
7. `(User application dir)/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib`

4~7 are fallbacks on `xcode-select -p` pointing Command Line Tools OS X for Xcode, because that does not contains `sourcekitd.framework`.  

#### Specifying toolchain dir on launching `sourcekitten`
```sh
TOOLCHAIN_DIR="/Library/Developer/Toolchains/swift-latest.xctoolchain/" sourcekitten …
# or
DEVELOPER_DIR="/Applications/Xcode-beta.app" sourcekitten … # `xcode-select -p` is affected by that
```

norio-nomura commented:
Added tag `0.10.1-test` for testing on SwiftLint

norio-nomura commented:
Change test tag from `0.10.1-test` to `0.11.0-test` for keep compatibility to existing dependent.

jpsim commented on Source/SourceKittenFramework/Loader.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+let toolchainLoader = Loader(pathes: [
```
The plural of path is paths.

jpsim commented on Source/SourceKittenFramework/Loader.swift:
```diff
@@ -0,0 +1,139 @@
+    var end = output.startIndex
+    var contentsEnd = output.startIndex
+    output.getLineStart(&start, end: &end, contentsEnd: &contentsEnd, forRange: start..<start)
+    let xcodeSelectPath = output.substringWithRange(start..<contentsEnd)
```
`end` isnt used after this so you could just pass `nil` instead.

jpsim commented:
Excellent work here! This seems to be on the right track, but I'm sure we can avoid the swift interface generation and use the existing headers.

norio-nomura commented:
Thank you for reviews!

>  but I'm sure we can avoid the swift interface generation and use the existing headers.

Are there method generating them from header files instead of puting them manually in `SourceKitten/Source/SourceKittenFramework/generated_interfaces`?
Or did you mean another method?

zwaldowski commented on CHANGELOG.md:
```diff
@@ -2,7 +2,11 @@
+  runtime by `SourceKittenFramework`. Depending client might need some
+  modification if referencing them from language other than Swift.  
+  [Norio Nomura](https://github.com/norio-nomura)
+  [#167](https://github.com/jpsim/SourceKitten/issues/167)
```
Re: this limitation, the `Loader` could try to load plain `sourcekitd` and `libclang.dylib` (the unqualified paths) to inherit `DYLD_FRAMEWORK_PATH` and `DYLD_LIBRARY_PATH`. They'd have to handle the error from that.

norio-nomura replied:
Sorry, my lack of reading english.
I can't understand what you suggested.

zwaldowski replied:
If you at call `dlopen("libclang.dylib", …)` or `dlopen("sourcekit.framework/sourcekitd", …)` at some point, the normal paths will be searched. This would allow something that works by setting `DYLD_LIBRARY_PATH` and `DYLD_FRAMEWORK_PATH` (like Jazzy) to continue to work with no changes. This would require some changes to the code that's there right now, such as handling the error from `dlopen`.

norio-nomura replied:
Thanks! I understand.
It sounds nice idea. :+1:
I'll try it.

zwaldowski commented on Source/SourceKittenFramework/create_swift_from_generated_interface.py:
```diff
@@ -0,0 +1,119 @@
+r"""public func \1\2 {
+    return _\1\2
+}
+private let _\1: @convention(c) \2 = library.loadSymbol("\1")
```
Have you profiled this PR? This technique may slow down startup; each of these global-level `lets` synthesizes another `dispatch_once`.

Additionally, why is the declaration `public`? Why would we want to expose these functions outside SourceKittenFramework?

norio-nomura replied:
I profiled it again.
On `sourcekitten doc', `sourcekitten syntax` and `sourcekitten structure`, `DynamicLinkLibrary`'s code takes less than 2ms.
On 'swiftlint lint', `DynamicLinkLibrary`'s code takes less than 2ms.

> why is the declaration public?

If clients directly calling `sourcekitd` or ` clang` exists, they can achieve them using SourceKittenFramework's wrapper.
I don't know whether such client exists or not.

norio-nomura commented:
Updated fail over logic in `Loader.load()`

- Try loading next existing library on `dlopen()` failed
- Try loading by simple relative path after all failed library loading by existing full paths. https://github.com/jpsim/SourceKitten/pull/176#discussion_r53730536

jpsim commented on Source/SourceKittenFramework/Loader.swift:
```diff
@@ -0,0 +1,145 @@
+            }
+        }
+
+        // try loading with simple path that depends resovling to DYLD
```
resolving

norio-nomura replied:
Thanks!

norio-nomura commented on Source/SourceKittenFramework/Loader.swift:
```diff
@@ -0,0 +1,145 @@
+/// `launch-with-toolchain` set toolchain path to "XCODE_DEFAULT_TOOLCHAIN_OVERRIDE" environment
+/// variable.
+private let xcodeDefaultToolchainOverride: String? =
+    NSProcessInfo.processInfo().environment["XCODE_DEFAULT_TOOLCHAIN_OVERRIDE"]
```
Note: I'v found that Xcode 7.3 beta 4 (7D152p) does not have `launch-with-toolchain`.
```sh
% DEVELOPER_DIR="/Applications/Xcode-beta.app" xcrun launch-with-toolchain
2016-02-23 15:35:53.988 xcodebuild[95151:1105215] [MT] PluginLoading: Required plug-in compatibility UUID E71C2CFE-BFD8-4044-8F06-00AE685A406C for plug-in at path '~/Library/Application Support/Developer/Shared/Xcode/Plug-ins/ExtraBuildPhase.xcplugin' not present in DVTPlugInCompatibilityUUIDs
xcrun: error: unable to find utility "launch-with-toolchain", not a developer tool or in PATH
```

norio-nomura commented:
> but I'm sure we can avoid the swift interface generation and use the existing headers.

Oh, I tested #95 and it worked.
@jpsim Did you mean that?

norio-nomura commented:
> Oh, I tested #95 and it worked.

After some investigation, I found that generated Swift interface by #95 does not contain `func` and can't be used in this PR instead of `SourceKitten/Source/SourceKittenFramework/generated_interfaces`. :disappointed:

jpsim commented:
> Oh, I tested #95 and it worked.
> @jpsim Did you mean that?

No, I meant why aren't we using `.h` headers for symbol resolution? As far as I can tell, that's the purpose of the generated interfaces you've introduced here.

norio-nomura commented:
Sorry, my bad naming.  Files in `generated_interfaces` are seeds of `created_*.swift` feeding to `create_swift_from_generated_interface.py` script. Those are never compiled as swift source. We are still using `.h`.

jpsim commented:
Why are the Swift files necessary at all?

norio-nomura commented:
`created_*.swift` supply wrapper `func`s of `sourcekitd` and `libclang` API. 
`create_swift_from_generated_interface.py` and `generated_interface/*.txt` are used to creating those wrapper. I thought those should be contained into repository for recording how those wrapper were created.

norio-nomura commented:
I'll try to add some documentation about that into the repository, not in here.

jpsim commented:
Yes please, I still don't understand why that's required.

norio-nomura commented:
I reorganized files and added documentation.
https://github.com/jpsim/SourceKitten/blob/dfbb42880809eeea9f19b90e0720e3c506d69e4a/Source/SourceKittenFramework/library_wrapper/README.md

norio-nomura commented:
Now, I can produce Generated Swift Interface from headers by using small swift program calling `sourcekitd`. I will change this PR to use that for creating `library_wrapper_*.swift` instead of adding Generated Swift Interface as .txt files to repository.
On that process, I want to remove following:
```objective-c
#import <SourceKittenFramework/Index.h>
#import <SourceKittenFramework/Documentation.h>
#import <SourceKittenFramework/CXCompilationDatabase.h>
#import <SourceKittenFramework/sourcekitd.h>
```
from `SourceKittenFramework.h`.
I recognize that those are used for instead of bridging header.
Instead of using `#import` in the header, `import SourceKit` and `import Clang_C` will work on building by Xcode. That will remove `#if SWIFT_PACKAGE` on importing them.

norio-nomura commented:
> I want to remove following:
> ```swift
> #import <SourceKittenFramework/Index.h>
> #import <SourceKittenFramework/Documentation.h>
> #import <SourceKittenFramework/CXCompilationDatabase.h>
> #import <SourceKittenFramework/sourcekitd.h>
> ```
> from SourceKittenFramework.h.

I gave up that because making them require each clients of SourceKittenFramework to importing `SourceKit` and `Clang_C` separately.
So, I'll push without that.

norio-nomura commented:
Removed tag `0.11.0-test`

norio-nomura commented:
Updated.
- Added `interfacegen` that generates Swift interface from header files
- Stopped containing manually created Generated Swift Interface files (.txt).

norio-nomura commented:
Rebased.

norio-nomura commented:
This is replaced by #185
